spec/lib/search_spec.rb

```diff
@@ -609,49 +610,93 @@ def create_pm(users:, group: nil)
-        [
-          "@#{participant.username} in:personal-direct",
-          "@#{participant.username} iN:pErSoNaL-dIrEcT",
-        ].each do |query|
-          results = Search.execute(query, guardian: Guardian.new(current))
-          expect(results.posts.size).to eq(2)
-          expect(results.posts.map(&:topic_id)).to contain_exactly(pm_2.id, pm.id)
-          expect(results.posts.map(&:user_id).uniq).to eq([participant.id])
+        it "doesn't include PMs that have more than 2 participants" do
+          _pm = create_pm(users: [current, participant, participant_2])
+          results = Search.execute("@#{participant.username} in:personal-direct", guardian: Guardian.new(current))
+          expect(results.posts.size).to eq(0)
         end
 
-        results = Search.execute("@me in:personal-direct", guardian: Guardian.new(current))
-        expect(results.posts.size).to eq(3)
-        expect(results.posts.map(&:topic_id)).to contain_exactly(pm_3.id, pm_2.id, pm.id)
-        expect(results.posts.map(&:user_id).uniq).to eq([current.id])
+        it "doesn't include PMs that have groups" do
+          _pm = create_pm(users: [current, participant], group: group)
+          results = Search.execute("@#{participant.username} in:personal-direct", guardian: Guardian.new(current))
+          expect(results.posts.size).to eq(0)
+        end
       end
 
-      it "doesn't include PMs that have more than 2 participants" do
-        _pm = create_pm(users: [current, participant, participant_2])
-        results = Search.execute("@#{participant.username} in:personal-direct", guardian: Guardian.new(current))
-        expect(results.posts.size).to eq(0)
-      end
+      context "group_messages flag" do
+        it 'returns results correctly for a PM in a given group' do
+          pm = create_pm(users: [participant, participant_2], group: group)
+
+          results = Search.execute("group_messages:#{group.name}", guardian: Guardian.new(current))
+          expect(results.posts).to contain_exactly(pm.first_post)
+
+          results = Search.execute("secret group_messages:#{group.name}", guardian: Guardian.new(current))
+          expect(results.posts).to contain_exactly(pm.first_post)
+        end
 
-      it "doesn't include PMs that have groups" do
-        _pm = create_pm(users: [current, participant], group: group)
-        results = Search.execute("@#{participant.username} in:personal-direct", guardian: Guardian.new(current))
-        expect(results.posts.size).to eq(0)
+        it 'returns nothing if user is not a group member' do
+          pm = create_pm(users: [current, participant], group: group)
+
+          results = Search.execute("group_messages:#{group.name}", guardian: Guardian.new(non_participant))
+          expect(results.posts.size).to eq(0)
+
+          # even for admins
+          results = Search.execute("group_messages:#{group.name}", guardian: Guardian.new(admin))
+          expect(results.posts.size).to eq(0)
```

Comment:
Do we limit admin access like this already in other areas?

Reply:
We don't specifically limit it, but the `group:` keyword search works the same, it only returns results if the user is a gruop member.